File: _x/talks/objectively_harmful/code/php/b_interface.php
<?php

// BGN1 OMIT
interface greeter { public function greeting($str); }

function meet($name, greeter ...$greeters) {
    foreach ($greeters as $greeter) {
		echo $greeter->greeting($name)."\n";
	}
}

class Human implements greeter {
    // ...
	protected $name; // OMIT
    function __construct($name) { $this->name = $name; } // OMIT
    // OMIT
    public function greeting($name) { return "Hello, $name. I'm $this->name."; } // OMIT
}

class Wolf implements greeter {
    // ...
    protected $freq = 1; // OMIT
    function __construct($freq) { $this->freq = $freq; } // OMIT
    // OMIT
    public function greeting($_) { // OMIT
        $msg = str_repeat("woof ", $this->freq); // OMIT
        return trim($msg)."!"; // OMIT
    } // OMIT
}

$a = new Human("Alice");
$b = new Wolf(3);
meet("Dan", $a, $b);
// END1 OMIT

File: _x/talks/objectively_harmful/code/go/b_interface.go
package main

import (
	"fmt"
	"strings"
)

type human struct {
	name string
}

func (h human) greeting(name string) string {
	return fmt.Sprintf("Hello, %s. I'm %s.", name, h.name)
}

type wolf struct {
	freq int
}

func (w wolf) greeting(string) string {
	msg := strings.Repeat("woof ", w.freq)
	return fmt.Sprintf("%s!", strings.TrimSpace(msg))
}

// BGN1 OMIT
type greeter interface {
	greeting(name string) string
}

func meet(name string, gs ...greeter) {
	for _, g := range gs {
		fmt.Println(g.greeting(name))
	}
}

func main() {
	a := human{"Alice"}
	b := wolf{3}
	meet("Dan", a, b)
}

// END1 OMIT
